### ISSUE TITLE
UbiquitousComputing - If the token for `Vesting` is BNB, re-entrancy will allow double-claiming, draining the contract

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-03-zap-protocol-judging#5]
UbiquitousComputing

high

# If the token for `Vesting` is BNB, re-entrancy will allow double-claiming, draining the contract

## Summary

The function `Vesting.claim()` is used for the user to claim their vesting amount. However, in case the token to be vested is BNB, an external call will be made to the sender.

Since the function doesn't follow the checks-effects-interaction pattern, nor does it have any re-entrancy protection, it opens up room for re-entrancy attack.

This report describes the attack that will allow double-claiming

## Vulnerability Detail

When calling `Vesting.claim`, the function calculates the amount to be claimed by calculating claimable amounts through vesting checkpoints, adding the amount when it is claimable. The calculation starts at the previously claimed checkpoint.

```solidity
uint256 pctAmount;
uint256 i = s.index;
for (i; i <= vestingPoints.length - 1; i++) {
    if (block.timestamp >= vestingPoints[i][0]) {
        pctAmount += (s.userDeposit * vestingPoints[i][1]) / 10000;
    } else {
        break;
    }
}
```

After calculating the received amount, the amount will be transferred to the claiming user.

```solidity
if (pctAmount != 0) {
    if (address(token) == address(1)) {
        (bool sent, ) = payable(sender).call{value: pctAmount}(""); // @audit external call here
        require(sent, "Failed to send BNB to receiver");
    } else {
        token.safeTransfer(sender, pctAmount);
    }
    s.index = uint128(i);
    s.amountClaimed += pctAmount;
}
```

Note that the claiming part performs an external call to the sender before updating the state (specifically, the `s.index`). This opens up a re-entrancy attack that allows a checkpoint to be claimed twice. 

### Proof of concept

Alice is a Zap user, who is eligible for some BNB vesting. She uses a smart contract wallet as her vesting address. 

Assuming the contract has a single vesting point, is now claimable, and the amount for this vesting checkpoint for Alice is 1 BNB. The following attack will allow Alice to claim 2 BNB:
1. Note the initial state of `userdetails[ALICE]` (she hasn't claimed anything):
  - `s.index = 0`
  - `s.claimedAmount = 0 BNB`
2. Alice calls `claim()` normally. From lines 67-81, the contract calculates Alice's claimable amount, from checkpoint 0 to checkpoint 1, to be 1 BNB.
  - `s.index = 0`
  - `s.claimedAmount = 0 BNB`
3. At line 84, the contract sends 1 BNB using `call` to Alice. At this point:
  - `s.index = 0`
  - `s.claimedAmount = 0 BNB` 
  - They won't be updated until lines 89 and 90.
4. Alice receives 1 BNB, and her wallet's `receive()` function triggers `Vesting.claim()` again:
  - Since `s.index = 0`, the contract calculates claimable balance starting from 0. Step 2 and 3 are repeated, sending Alice another 1 BNB.
5. Alice is sent a total of 2 BNB. Final state:
  - `s.index = 1`
  - `s.claimedAmount = 2 BNB` 

The result is that checkpoint 0 is claimed twice, while Alice should've been able to claim it only once.

Alice's wallet can repeat step 4 for as long as it likes, until the contract is drained completely.

## Impact

BNB in `Vesting` can be drained by multiple-claiming using re-entrancy

## Code Snippet

https://github.com/sherlock-audit/2024-03-zap-protocol/blob/main/zap-contracts-labs/contracts/Vesting.sol#L67-L92

## Tool used

Manual Review

## Recommendation

Add a re-entrancy guard to the function, or move lines 89-90 before the external BNB send, following the CEI pattern.

Duplicate of #157
